tap_aggregator/src/metrics.rs

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023-, Semiotic AI, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{net::SocketAddr, panic};
+
+use axum::{http::StatusCode, response::IntoResponse, routing::get, Router, Server};
+use futures_util::FutureExt;
+use jsonrpsee::tracing::error;
+use log::{debug, info};
+use prometheus::TextEncoder;
+
+async fn handler_metrics() -> (StatusCode, String) {
+    let metric_families = prometheus::gather();
+    let encoder = TextEncoder::new();
+
+    match encoder.encode_to_string(&metric_families) {
+        Ok(s) => (StatusCode::OK, s),
+        Err(e) => {
+            error!("Error encoding metrics: {}", e);
+            (
+                StatusCode::INTERNAL_SERVER_ERROR,
+                format!("Error encoding metrics: {}", e),
+            )
+        }
+    }
+}
+
+async fn handler_404() -> impl IntoResponse {
+    (StatusCode::NOT_FOUND, "404 Not Found")
+}
+
+async fn _run_server(port: u16) {
+    let app = Router::new().route("/metrics", get(handler_metrics));
+    let app = app.fallback(handler_404);
+    let addr = SocketAddr::from(([0, 0, 0, 0], port));
+    let server = Server::bind(&addr).serve(app.into_make_service());
+
+    info!("Metrics server listening on {}", addr);
+
+    let res = server.await;
+
+    debug!("Metrics server stopped");
+
+    if let Err(err) = res {
+        panic!("Metrics server error: {:#?}", err);
+    };
+}
+
+pub async fn run_server(port: u16) {
+    // Code here is to abort program if there is a panic in _run_server
+    // Otherwise, when spawning the task, the panic will be silently ignored
+    let res = panic::AssertUnwindSafe(_run_server(port))
+        .catch_unwind()
+        .await;
```

Comment:
Is there a reason to panic and catch unwinds instead of just returning a result and handling it? `AssertUnwindSafe` appears to bypass compiler checks and could cause lead to issues eventually.

Reply:
This is the only way to have a failure in a spawned thread immediately stop the execution of the program. Otherwise, errors and panics are silent.
I was inspired by https://github.com/graphprotocol/graph-node/blob/ba5f12f346f0138f09454dc719095711e7d71cf9/graph/src/task_spawn.rs#L18-L31, and all the solutions I've found on the internet were pointing to that same solution.

Reply:
I'd also that even if I handled errors gracefully, using inter-thread signals for example, panics in the spawned thread would still be silent unless caught using this method.